File: apps/macos/Kelpie/Account/UOAAuthorization.swift
import CryptoKit
import Foundation
import Security

struct UOAAuthorization {
    static let callback = "com.unlikeotherai.kelpie://oauth/callback"
    static let scopes = "openid profile settings.read settings.write"
    let state: String
    let verifier: String

    init() throws {
        state = try Self.randomValue()
        verifier = try Self.randomValue()
    }

    init(state: String, verifier: String) {
        self.state = state
        self.verifier = verifier
    }

    var challenge: String { Self.base64URL(Data(SHA256.hash(data: Data(verifier.utf8)))) }

    func url(clientID: String) throws -> URL {
        guard var components = URLComponents(string: UOATransport.origin + "/oauth/authorize") else {
            throw UOATransport.Failure(status: 400)
        }
        components.queryItems = [
            URLQueryItem(name: "response_type", value: "code"),
            URLQueryItem(name: "client_id", value: clientID),
            URLQueryItem(name: "redirect_uri", value: Self.callback),
            URLQueryItem(name: "state", value: state),
            URLQueryItem(name: "scope", value: Self.scopes),
            URLQueryItem(name: "code_challenge", value: challenge),
            URLQueryItem(name: "code_challenge_method", value: "S256")
        ]
        guard let url = components.url else { throw UOATransport.Failure(status: 400) }
        return url
    }

    func code(from url: URL) throws -> String {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              components.scheme == "com.unlikeotherai.kelpie", components.host == "oauth",
              components.path == "/callback", components.user == nil, components.password == nil,
              components.port == nil, components.fragment == nil else { throw UOATransport.Failure(status: 401) }
        let items = components.queryItems ?? []
        guard items.filter({ $0.name == "state" }).count == 1,
              items.first(where: { $0.name == "state" })?.value == state,
              items.filter({ $0.name == "code" }).count == 1,
              !items.contains(where: { $0.name == "error" }),
              let code = items.first(where: { $0.name == "code" })?.value, !code.isEmpty else {
            throw UOATransport.Failure(status: 401)
        }
        return code
    }

    private static func randomValue() throws -> String {
        var bytes = [UInt8](repeating: 0, count: 32)
        guard SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess else {
            throw UOATransport.Failure(status: 0)
        }
        return base64URL(Data(bytes))
    }

    private static func base64URL(_ data: Data) -> String {
        data.base64EncodedString().replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "")
    }
}

File: apps/android/app/src/main/java/com/kelpie/browser/account/UOAAuthorization.kt
package com.kelpie.browser.account

import java.net.URI
import java.net.URLDecoder
import java.net.URLEncoder
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64

/** Public native OAuth uses S256 PKCE and a one-use exact callback. */
class UOAAuthorization(
    val state: String = randomValue(),
    val verifier: String = randomValue(),
) {
    val challenge: String = base64(MessageDigest.getInstance("SHA-256").digest(verifier.toByteArray(Charsets.US_ASCII)))

    fun url(clientId: String): String {
        val fields =
            linkedMapOf(
                "response_type" to "code",
                "client_id" to clientId,
                "redirect_uri" to CALLBACK,
                "scope" to SCOPES,
                "state" to state,
                "code_challenge" to challenge,
                "code_challenge_method" to "S256",
            )
        return UOATransport.ORIGIN + "/oauth/authorize?" + fields.entries.joinToString("&") { "${encode(it.key)}=${encode(it.value)}" }
    }

    private fun fields(callback: String): List<Pair<String, String>> {
        val uri = URI(callback)
        require(
            uri.scheme == "com.unlikeotherai.kelpie" &&
                uri.host == "oauth" &&
                uri.path == "/callback" &&
                uri.userInfo == null &&
                uri.port == -1 &&
                uri.fragment == null,
        ) { "Invalid login callback" }
        val fields =
            uri.rawQuery.orEmpty().split('&').map { item ->
                val pair = item.split('=', limit = 2)
                decode(pair[0]) to decode(pair.getOrElse(1) { "" })
            }
        require(fields.count { it.first == "state" } == 1 && fields.single { it.first == "state" }.second == state) { "Invalid login response" }
        return fields
    }

    fun denied(callback: String): Boolean = fields(callback).any { it.first == "error" }

    fun code(callback: String): String {
        val fields = fields(callback)
        require(fields.count { it.first == "code" } == 1 && fields.none { it.first == "error" }) { "Invalid login response" }
        return fields.single { it.first == "code" }.second.also { require(it.isNotBlank()) }
    }

    companion object {
        const val CALLBACK = "com.unlikeotherai.kelpie://oauth/callback"
        const val SCOPES = "openid profile settings.read settings.write"

        private fun randomValue(): String = base64(ByteArray(32).also(SecureRandom()::nextBytes))

        private fun base64(bytes: ByteArray): String = Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)

        private fun encode(value: String) = URLEncoder.encode(value, "UTF-8")

        private fun decode(value: String) = URLDecoder.decode(value, "UTF-8")
    }
}
